### DISCUSSION TITLE
Yak Rover LLM

### DISCUSSION THREAD

**Discussion by rhettg** [YakRoboticsGarage/YARG#8] [Show and tell]
I wired up an GPT-4V application to try to run my rover on a simple task like "navigate over to the baby Yoda doll".

![iTerm2 Jm3XWE](https://github.com/YakRoboticsGarage/YARG/assets/50074/8e593ade-87b5-4ae2-a3c9-15237d85974a)

Video Demo Here: https://youtu.be/SVGcx4AvwFs

This is based on [Batteries Not Included](https://github.com/rhettg/batteries) and the code is here: https://github.com/rhettg/batteries/tree/main/yakllm

Some failure modes I encountered:

* Agent-to-Agent design (Rover "talks" to Eyes) is failure prone. Sometimes the Rover hallucinates an answer which causes the Eyes to hallucinate.
  * I'm using this architecture because I"m using two models: `gpt-4-turbo` for the rover and `gpt-4-vision-preview` for the eyes. Because, from [here](https://platform.openai.com/docs/guides/vision): "Currently, GPT-4 Turbo with vision does not support the message.name parameter, functions/tools, response_format parameter, and we currently set a low max_tokens default which you can override."
  * It's possible more straightforward function call architecture could work better. I should try it.
* Eyes show pretty poor accuracy when it comes to positioning objects in a scene. This is largely expected, but it was good to prove it out.
  * From OpenAI: "The model is best at answering general questions about what is present in the images. While it does understand the relationship between objects in images, it is not yet optimized to answer detailed questions about the location of certain objects in an image."

But overall it's pretty awesome to watch the ai drive the rover around, even if it does a terrible job at it.

**Comment by rhettg** [comment]
Ok, well I asked ChatGPT about an image 4 times and it got a much better answer every time than I have been getting from my implementation. So... we might have an implementation or prompt issue.

<img width="859" alt="image" src="https://github.com/YakRoboticsGarage/YARG/assets/50074/13119e87-e842-460d-8780-346a83868371">

**Reply by ic** [reply]
This is currently what really makes "engineering" with an LLM component difficult... No tip to share on this one. In similar context I have tried using a [vector DB](https://github.com/chroma-core/chroma) to create equivalence between questions that should be (so what time is it? and I wonder what the time is get in the same basket), but it is a lot of work (daunting if necessary). Better luck with "[guardrails](https://github.com/NVIDIA/NeMo-Guardrails)", but it seems a bit off for the problem at hand.
